core/modfunc.go

```diff
@@ -0,0 +1,323 @@
+	objDef *ObjectTypeDef,
+	runtime *Container,
+	metadata *Function,
+) (*ModuleFunction, error) {
+	returnType, ok, err := mod.ModTypeFor(ctx, metadata.ReturnType, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get mod type for function %q return type: %w", metadata.Name, err)
+	}
+	if !ok {
+		return nil, fmt.Errorf("failed to find mod type for function %q return type: %q", metadata.Name, metadata.ReturnType.ToType())
+	}
+
+	argTypes := make(map[string]*UserModFunctionArg, len(metadata.Args))
+	for _, argMetadata := range metadata.Args {
+		argModType, ok, err := mod.ModTypeFor(ctx, argMetadata.TypeDef, true)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get mod type for function %q arg %q type: %w", metadata.Name, argMetadata.Name, err)
+		}
+		if !ok {
+			return nil, fmt.Errorf("failed to find mod type for function %q arg %q type", metadata.Name, argMetadata.Name)
+		}
+		argTypes[argMetadata.Name] = &UserModFunctionArg{
+			metadata: argMetadata,
+			modType:  argModType,
+		}
+	}
+
+	return &ModuleFunction{
+		root:       root,
+		mod:        mod,
+		modID:      modID,
+		objDef:     objDef,
+		runtime:    runtime,
+		metadata:   metadata,
+		returnType: returnType,
+		args:       argTypes,
+	}, nil
+}
+
+type CallOpts struct {
+	Inputs         []CallInput
+	ParentVal      map[string]any
+	Cache          bool
+	Pipeline       pipeline.Path
+	SkipSelfSchema bool
+}
+
+type CallInput struct {
+	Name  string
+	Value dagql.Typed
+}
+
+func (fn *ModuleFunction) Call(ctx context.Context, caller *idproto.ID, opts *CallOpts) (t dagql.Typed, rerr error) {
+	mod := fn.mod
+
+	lg := bklog.G(ctx).WithField("module", mod.Name()).WithField("function", fn.metadata.Name)
+	if fn.objDef != nil {
+		lg = lg.WithField("object", fn.objDef.Name)
+	}
+	ctx = bklog.WithLogger(ctx, lg)
+
+	callInputs := make([]*FunctionCallArgValue, len(opts.Inputs))
+	hasArg := map[string]bool{}
+	for i, input := range opts.Inputs {
+		normalizedName := gqlArgName(input.Name)
+		arg, ok := fn.args[normalizedName]
+		if !ok {
+			return nil, fmt.Errorf("failed to find arg %q", input.Name)
+		}
+
+		name := arg.metadata.OriginalName
+
+		converted, err := arg.modType.ConvertToSDKInput(ctx, input.Value)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert arg %q: %w", input.Name, err)
+		}
+
+		encoded, err := json.Marshal(converted)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal arg %q: %w", input.Name, err)
+		}
+
+		callInputs[i] = &FunctionCallArgValue{
+			Name:  name,
+			Value: encoded,
+		}
+
+		hasArg[name] = true
+	}
+
+	for _, arg := range fn.metadata.Args {
+		name := arg.OriginalName
+		if hasArg[name] || arg.DefaultValue == nil {
+			continue
+		}
+		callInputs = append(callInputs, &FunctionCallArgValue{
+			Name:  name,
+			Value: arg.DefaultValue,
+		})
+	}
+
+	callerDigestInputs := []string{}
+	{
+		callerIDDigest, err := caller.Digest() // FIXME(vito) canonicalize, once all that's implemented
```

Comment:
Kind of - there's a prospective `(*ID).Canonical()` implementation but none of the fields set `Meta()` yet so nothing's impacted either way. I've also noticed `Canonical()` is pretty slow, so I'm wondering if we need a faster way to calculate it. Which also ties into whether/when we make IDs smaller by deduping them into a `map[digest]ID` where ID refers to its parent/arg/module IDs by digest instead of by embedding them, and the trade-offs involved. So it's a bit of a rabbit hole. :sweat_smile: 

More broadly I think there's a bit more to figure out with the whole 'canonicalizing IDs' mechanism - like whether/when it even makes sense to use a _non_-canonicalized ID, and whether loading an ID should attempt a cache store lookup against the canonicalized digest instead of just evaluating it and letting each selection hit the cache.

I've been punting on this for now since the scope is pretty big already and the only immediate implication is busting the query cache, and right now they're not cached at all. Would definitely want to address all this sooner rather than later though.